driver/src/contracts/batched_auction_data_reader.rs

```diff
@@ -0,0 +1,255 @@
+use super::stablex_auction_element::{StableXAuctionElement, AUCTION_ELEMENT_WIDTH};
+use dfusion_core::models::{AccountState, BatchInformation, Order};
+use std::collections::HashMap;
+use web3::types::{H160, U256};
+
+/// Handles reading of auction data that has been encoded with the smart
+/// contract's `encodeAuctionElement` function.
+pub struct BatchedAuctionDataReader {
+    index: U256,
+    pub account_state: AccountState,
+    /// All orders in the order they were received.
+    pub orders: Vec<Order>,
+    /// The total number of orders per user.
+    pub user_order_counts: HashMap<H160, u16>,
+    /// Used when reading data from the smart contract in batches.
+    pub pagination: Pagination,
+}
+
+/// Data for the next call to the smart contract's `getEncodedUsersPaginated`
+/// function (which should be named `getEncodedOrdersPaginated`).
+pub struct Pagination {
+    pub previous_page_user: H160,
+    pub previous_page_user_offset: u64,
+}
+
+impl BatchedAuctionDataReader {
+    pub fn new(index: U256) -> BatchedAuctionDataReader {
+        BatchedAuctionDataReader {
+            index,
+            account_state: AccountState::default(),
+            orders: Vec::new(),
+            user_order_counts: HashMap::new(),
+            pagination: Pagination {
+                previous_page_user: H160::zero(),
+                previous_page_user_offset: 0,
+            },
+        }
+    }
+
+    /// Applies one batch of data.
+    ///
+    /// A batch can come from `getEncodedUsersPaginated` or `getEncodedOrders`.
+    /// In the latter case there is only one batch.
+    ///
+    /// Returns the number of orders in the data.
+    ///
+    /// Panics if `packed_auction_bytes` is not a multiple of
+    /// `AUCTION_ELEMENT_WIDTH`.
+    pub fn apply_batch(&mut self, packed_auction_bytes: &[u8]) -> u64 {
+        let previous_order_count = self.orders.len();
+        self.apply_auction_data(&packed_auction_bytes);
+        let number_of_added_orders = (self.orders.len() - previous_order_count) as u64;
+        if number_of_added_orders == 0 {
+            return 0;
+        }
+        let last_order_user = self
+            .orders
+            .last()
+            .expect("have at least one order")
+            .account_id;
+        if self.pagination.previous_page_user == last_order_user {
+            self.pagination.previous_page_user_offset += number_of_added_orders;
+        } else {
+            self.pagination.previous_page_user = last_order_user;
+            self.pagination.previous_page_user_offset = number_of_added_orders;
```

Comment:
Is this correct? I believe this should be the number of orders of this particular user. Consider a page like: `user1, user2, user2`

The offset should be 2 but I believe here it would be 1. Maybe we can reuse user_order_counts logic here?